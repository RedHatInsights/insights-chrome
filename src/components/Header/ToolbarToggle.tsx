import React, { useState } from 'react';
import { useAtomValue } from 'jotai';
import { Dropdown, DropdownItem, DropdownList } from '@patternfly/react-core/dist/dynamic/components/Dropdown';
import { MenuToggle } from '@patternfly/react-core/dist/dynamic/components/MenuToggle';
import { PopoverPosition } from '@patternfly/react-core/dist/dynamic/components/Popover';

import ChromeLink from '../ChromeLink/ChromeLink';
import { isPreviewAtom } from '../../state/atoms/releaseAtom';

export type ToolbarToggleDropdownItem = {
  url?: string;
  appId?: string;
  target?: string;
  title: string;
  onClick?: (event: MouseEvent | React.MouseEvent<any, MouseEvent> | React.KeyboardEvent<Element>) => void;
  isHidden?: boolean;
  isDisabled?: boolean;
  rel?: string;
};

export type ToolbarToggleProps = {
  icon?: React.ElementType;
  dropdownItems: ToolbarToggleDropdownItem[];
  widgetType?: string | number;
  className?: string;
  id?: string | number;
  hasToggleIndicator?: null;
  ouiaId?: string;
  isHidden?: boolean;
  ariaLabel?: string;
};

const ToolbarToggle = (props: ToolbarToggleProps) => {
  const [isOpen, setIsOpen] = useState(false);
  const isPreview = useAtomValue(isPreviewAtom);

  const onSelect = () => {
    setIsOpen((prev) => !prev);
  };

  const onToggle = () => setIsOpen((prev) => !prev);

  const onClickInternal = (
    ev: MouseEvent | React.KeyboardEvent<Element> | React.MouseEvent<any, MouseEvent>,
    url?: string,
    onClick?: ToolbarToggleDropdownItem['onClick']
  ) => {
    ev.preventDefault();
    if (url) {
      window.location.href = `${url}`;
    }

    if (onClick) {
      onClick(ev);
    }
  };

  // Render the question mark icon items
  const dropdownItems = props.dropdownItems.map(
    ({ url, appId, title, onClick, isHidden, isDisabled, target = '_blank', rel = 'noopener noreferrer', ...rest }) =>
      !isHidden ? (
        <DropdownItem
          key={title}
          ouiaId={title}
          isDisabled={isDisabled}
          component={
            appId && url
              ? ({ className: itemClassName }) => (
                  <ChromeLink {...rest} className={itemClassName} href={url} target={target} rel={rel} isBeta={isPreview} appId={appId}>
                    {title}
                  </ChromeLink>
                )
              : url
              ? 'a'
              : 'button'
          }
          // Because the urls are using 'a', don't use onClick for accessibility
          // If it is a button, use the onClick prop
          {...(appId
            ? {}
            : url
            ? {
                href: url,
                target,
                rel,
                ...rest,
              }
            : { onClick: (ev) => onClickInternal(ev, url, onClick) })}
        >
          {title}
        </DropdownItem>
      ) : (
        <React.Fragment key="fragment" />
      )
  );

  return (
    <Dropdown
      popperProps={{
        position: PopoverPosition.right,
      }}
      onOpenChange={setIsOpen}
      toggle={(toggleRef) => (
        <MenuToggle
          ref={toggleRef}
          variant={props.icon ? 'plain' : 'default'}
          className={props.className}
          id={props.id?.toString()}
          onClick={onToggle}
          aria-label={props.ariaLabel}
          isExpanded={isOpen}
          icon={props.icon && <props.icon />}
        />
      )}
      isOpen={isOpen}
      onSelect={onSelect}
      ouiaId={props.ouiaId}
    >
      <DropdownList>{dropdownItems}</DropdownList>
    </Dropdown>
  );
};

export default ToolbarToggle;
