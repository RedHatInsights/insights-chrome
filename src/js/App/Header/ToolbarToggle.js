import React, { Component } from 'react';

import { DropdownToggle, DropdownItem, DropdownPosition, Dropdown } from '@patternfly/react-core';
import PropTypes from 'prop-types';
import ChromeLink from '../Sidenav/Navigation/ChromeLink';
import { isBeta } from '../../utils';

class ToolbarToggle extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isOpen: false,
    };
    this.onSelect = this.onSelect.bind(this);
    this.onToggle = this.onToggle.bind(this);
  }

  onSelect() {
    this.setState({ isOpen: !this.state.isOpen });
  }

  onToggle(isOpen) {
    this.setState({ isOpen });
  }

  onClick(ev, url, onClick) {
    ev.preventDefault();
    if (url) {
      window.location.href = `${url}`;
    }

    if (onClick) {
      onClick(ev);
    }
  }

  render() {
    // Render the questionmark icon items
    const dropdownItems = this.props.dropdownItems.map(
      ({ url, appId, title, onClick, isHidden, target = '_blank', rel = 'noopener noreferrer', ...rest }) =>
        !isHidden ? (
          <DropdownItem
            key={title}
            ouiaId={title}
            component={
              appId ? (
                <ChromeLink href={url} target={target} rel={rel} isBeta={isBeta()} appId={appId}>
                  {title}
                </ChromeLink>
              ) : url ? (
                'a'
              ) : (
                'button'
              )
            }
            // Because the urls are using 'a', don't use onClick for accessibility
            // If it is a button, use the onClick prop
            {...(appId
              ? {}
              : url
              ? {
                  href: url,
                  target,
                  rel,
                  ...rest,
                }
              : { onClick: (ev) => this.onClick(ev, url, onClick) })}
          >
            {title}
          </DropdownItem>
        ) : (
          <React.Fragment key="fragment" />
        )
    );

    const toggle = (
      <DropdownToggle
        className={this.props.className}
        id={this.props.id}
        ouiaId={this.props.id}
        toggleIndicator={this.props.hasToggleIndicator}
        onToggle={this.onToggle}
      >
        {this.props.icon && <this.props.icon />}
      </DropdownToggle>
    );

    return (
      <Dropdown
        position={DropdownPosition.right}
        toggle={toggle}
        isOpen={this.state.isOpen}
        dropdownItems={dropdownItems}
        onSelect={this.onSelect}
        ouiaId={this.props.ouiaId}
        isPlain
      />
    );
  }
}

ToolbarToggle.propTypes = {
  icon: PropTypes.elementType,
  dropdownItems: PropTypes.array,
  widgetType: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  className: PropTypes.string,
  id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  hasToggleIndicator: PropTypes.bool,
  ouiaId: PropTypes.string,
  isHidden: PropTypes.bool,
};

export default ToolbarToggle;
