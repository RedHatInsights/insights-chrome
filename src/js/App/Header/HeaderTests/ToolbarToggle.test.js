import React from 'react';
import toJson from 'enzyme-to-json';
import { shallow } from 'enzyme';
import ToolbarToggle from '../ToolbarToggle';

describe('ToolbarToggle', () => {

    it('should render correctly', () =>{
        const mockOnClick = jest.fn();
        const toolbarToggleProps = {
            dropdownItems: [
                {
                    url: 'url1',
                    title: 'title1',
                    onClick: mockOnClick
                }, {
                    title: 'title2',
                    onClick: mockOnClick
                }
            ],
            isOpen: true
        };
        const wrapper = shallow(<ToolbarToggle { ...toolbarToggleProps }/>);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

});
