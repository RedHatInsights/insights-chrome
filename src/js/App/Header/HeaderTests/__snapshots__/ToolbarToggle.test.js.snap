// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToolbarToggle should render correctly 1`] = `
<Dropdown
  aria-label="Settings"
  dropdownItems={
    Array [
      <DropdownItem
        component="a"
        href="url1"
        rel="noopener noreferrer"
        target="_blank"
      >
        title1
      </DropdownItem>,
      <DropdownItem
        component="button"
        onClick={[Function]}
      >
        title2
      </DropdownItem>,
    ]
  }
  isOpen={false}
  isPlain={true}
  onSelect={[Function]}
  position="right"
  toggle={
    <DropdownToggle
      iconComponent={null}
      onToggle={[Function]}
    >
      <UNDEFINED />
    </DropdownToggle>
  }
/>
`;
